net: lwm2m: Thread safety in the engine related files

otkulseng opened the pull request:
Added mutex locks to the lwm2m_rd_client and the lwm2m_registry to  
protect the global resources in the files.

Signed-off-by: Ola Tangen Kulseng <ola.kulseng@nordicsemi.no>

pdgendt commented on subsys/net/lib/lwm2m/lwm2m_rd_client.c:
```diff
@@ -261,6 +271,7 @@ static void sm_handle_failure_state(enum sm_engine_state sm_state)
 	} else if (client.engine_state == ENGINE_DEREGISTER_SENT) {
 		event = LWM2M_RD_CLIENT_EVENT_DEREGISTER_FAILURE;
 	}
+	k_mutex_unlock(&state_lock);
```
There is an early `return` within the locked context.

otkulseng replied:
thanks

rlubos commented on subsys/net/lib/lwm2m/lwm2m_rd_client.c:
```diff
@@ -1003,14 +1029,16 @@ static int sm_registration_done(void)
 	    (client.engine_state != ENGINE_REGISTRATION_DONE_RX_OFF) &&
 	    (((k_uptime_get() - client.last_tx) / 1000) >=
 	     CONFIG_LWM2M_QUEUE_MODE_UPTIME)) {
+		k_mutex_unlock(&state_lock);
 		set_sm_state(ENGINE_REGISTRATION_DONE_RX_OFF);
+		k_mutex_lock(&state_lock, K_FOREVER);
```
Why is the mutex unlocked here?

otkulseng replied:
It may not be a problem here, but the set_sm_state function calls the lwm2m_engine_context_close function from lwm2m_message_handling.c. This function uses the global resource messages that may also need a mutex. For safety, I tried to limit the number of function calls to other modules with mutexes.

SeppoTakalo commented on subsys/net/lib/lwm2m/lwm2m_rd_client.c:
```diff
@@ -373,13 +394,17 @@ int engine_trigger_bootstrap(void)
 #if defined(CONFIG_LWM2M_RD_CLIENT_SUPPORT_BOOTSTRAP)
 	if (!sm_is_registered()) {
 		/* Bootstrap is not possible to trig */
+		k_mutex_lock(&state_lock, K_FOREVER);
 		LOG_WRN("Cannot trigger bootstrap from state %u", client.engine_state);
+		k_mutex_unlock(&state_lock);
 		return -EPERM;
 	}
 
 	LOG_INF("Server Initiated Bootstrap");
 	client.use_bootstrap = true;
+	k_mutex_lock(&state_lock, K_FOREVER);
 	client.engine_state = ENGINE_INIT;
+	k_mutex_unlock(&state_lock);
 
```
I believe Zephyr mutexes are reentrant so we could lock multiple times, therefore this function and similar could be written like:
```suggestion
        k_mutex_lock(&state_lock, K_FOREVER);
	if (!sm_is_registered()) {
		/* Bootstrap is not possible to trig */
		LOG_WRN("Cannot trigger bootstrap from state %u", client.engine_state);
		k_mutex_unlock(&state_lock);
		return -EPERM;
	}

	LOG_INF("Server Initiated Bootstrap");
	client.use_bootstrap = true;
	client.engine_state = ENGINE_INIT;
	k_mutex_unlock(&state_lock);

```

SeppoTakalo commented on subsys/net/lib/lwm2m/lwm2m_registry.c:
```diff
@@ -40,6 +40,8 @@ LOG_MODULE_REGISTER(LOG_MODULE_NAME);
 #define BINDING_OPT_MAX_LEN 3 /* "UQ" */
 #define QUEUE_OPT_MAX_LEN   2 /* "Q" */
 
+static K_MUTEX_DEFINE(readwrite_lock);
+static K_MUTEX_DEFINE(registry_lock);
```
I don't understand, why we need two locks in one registry.
I think it should be fine to have both use the same mutex.

After all creating and deleting objects is a corner case that rarely happens. I don't see the point of optimizing it.

SeppoTakalo commented on subsys/net/lib/lwm2m/lwm2m_rd_client.c:
```diff
@@ -1375,7 +1408,9 @@ static int lwm2m_rd_client_init(const struct device *dev)
 {
 	client.ctx = NULL;
 	client.rd_message.ctx = NULL;
+	k_mutex_lock(&state_lock, K_FOREVER);
 	client.engine_state = ENGINE_IDLE;
+	k_mutex_unlock(&state_lock);
 	k_mutex_init(&client.mutex);
```
If client instance have mutex already, why we need another one for state variable, that is inside the client context?

I think we should use the client.mutex on all places.

otkulseng replied:
Agree

Commit:
net: lwm2m: Thread safe lwm2m registry

Created the mutex registry_lock to:
- protect read and write operations
- protect the registry.
Only partially finished as the functions like lwm2m_engine_get_obj
warrants a larger refactoring to completely thread safe the registry.

Signed-off-by: Ola Tangen Kulseng <ola.kulseng@nordicsemi.no>

Commit:
net: lwm2m: thread safety in the rd client

Used the mutex in the client to protect the state transitions of
client.state.

Signed-off-by: Ola Tangen Kulseng <ola.kulseng@nordicsemi.no>

rlubos commented:
@pdgendt Please revisit.
